### ISSUE TITLE
array.resize fails when using debugger with "ValueError: cannot resize an array that references or is referenced by another array in this way."

### ISSUE THREAD

**Issue by fabioz** [numpy/numpy#11538]
numpy version: 1.14.3
python version: 3.6.5

This was originally reported in https://github.com/Microsoft/ptvsd/issues/465 (error was on `numpy.loadtxt`, but the real issue lies in `resize` itself).

A sample to reproduce is:

```
import sys

def trace_dispatch(frame, event, arg):
    return trace_dispatch

sys.settrace(trace_dispatch)

def method():
    import numpy
    Y = numpy.array([1] * 10, numpy.float32)
    Y.resize([20, 20])

method()
```

When `Y.resize()` is reached, the following error is thrown:

```
ValueError: cannot resize an array that references or is referenced
by another array in this way.  Use the resize function.
```

So, it seems it's that way by design, but this also means that by design, any function which calls `array.resize` is not traceable (which I don't think is expected).

**Comment by seberg** [comment]
I doubt there is much we can do, this check is a safety hatch and it can be disabled with `checkref=False` (I am not sure if there are older versions that do not have the kwarg and always just resize). Unless there is an easy way to detect that this is a tracing situation and the extra references are related to tracing.

Note that as the arrays memory can be reassigned the refcheck is vital to avoid wrong usage of the `resize` method, which really is a very lowlevel call to the systems `realloc` and inherently very unsafe in python. Frankly, except for very limited and local usage in situations where you really need to preserve every little bit of RAM you have, I would probably just try to avoid it personally. (though with the `refcheck=True` default it should be safe on C-python)

**Comment by seberg** [comment]
I suppose the error could point to the `checkref` kwarg, but then it would also have to warn that it is very dangerous and should only be used with a decent knowledge about (python) memory management.

**Comment by fabioz** [comment]
Well, there's a way to check if it's from frame.f_locals.

You could check the case where you have 2 references (and not only 1) and do:

```
frame_locals = sys._getframe().f_locals
import gc
print(len(gc.get_referrers(Y)))
```

If the len of the referrers is still 2 after getting the locals (of the function which is calling array.resize), then it should be Ok.

Still, if this is not done, `numpy.loadtxt` should be changed to pass `refcheck=False` (or maybe `refcheck=TESTING_NUMPY_REFCHECK` and `TESTING_NUMPY_REFCHECK` could be True when running numpy tests but False by default for when it's being used by users).

**Comment by seberg** [comment]
Fair point, loadtxt knows what it is doing, so it should probably disable the `checkref` to avoid these problems, will rename the issue to that.
I am willing to believe that such a check can work, but am a bit wary about it without more though and tests.

**Comment by fabioz** [comment]
Seems a reasenable tradeof to me (although if it's really important, I'd say that it should be `True` when running numpy tests to make sure that some future contributor won't add a new reference to the array by mistake inside `numpy.loadtxt`).
